core-api/src/telemetry/metrics.rs

```diff
@@ -0,0 +1,250 @@
+    pub ids: HashSet<u64>,
+    /// If populated, these key values should also be used in addition to the referred-to
+    /// existing attributes when recording
+    pub new_attributes: Vec<MetricKeyValue>,
+}
+
+impl MetricAttributes {
+    /// Extend existing metrics attributes with others, returning a new instance
+    pub fn merge(&self, other: MetricAttributes) -> Self {
+        let mut me = self.clone();
+        match (&mut me, other) {
+            (MetricAttributes::OTel { ref mut kvs }, MetricAttributes::OTel { kvs: other_kvs }) => {
+                Arc::make_mut(kvs).extend((*other_kvs).clone());
+            }
+            (MetricAttributes::Lang(ref mut l), MetricAttributes::Lang(ol)) => {
+                l.ids.extend(ol.ids);
+                l.new_attributes.extend(ol.new_attributes);
+            }
+            _ => panic!("Cannot merge metric attributes of different kinds"),
+        }
+        me
+    }
+
+    /// Mutate self to add new kvs
+    pub fn add_new_attrs(&mut self, new_kvs: impl IntoIterator<Item = MetricKeyValue>) {
+        match self {
+            #[cfg(feature = "otel_impls")]
+            MetricAttributes::OTel { ref mut kvs, .. } => {
+                Arc::make_mut(kvs).extend(new_kvs.into_iter().map(Into::into));
+            }
+            MetricAttributes::Lang(ref mut attrs, ..) => {
+                attrs.new_attributes.extend(new_kvs.into_iter());
+            }
+        }
+    }
+}
+
+/// Options that are attached to metrics on a per-call basis
+#[derive(Clone, Debug, Default, derive_more::Constructor)]
+pub struct MetricsAttributesOptions {
+    pub attributes: Vec<MetricKeyValue>,
+}
+impl MetricsAttributesOptions {
+    pub fn extend(&mut self, new_kvs: impl IntoIterator<Item = MetricKeyValue>) {
+        self.attributes.extend(new_kvs.into_iter())
+    }
+}
+
+/// A K/V pair that can be used to label a specific recording of a metric
+#[derive(Clone, Debug)]
+pub struct MetricKeyValue {
+    pub key: String,
+    pub value: MetricValue,
+}
+impl MetricKeyValue {
+    pub fn new(key: impl Into<String>, value: impl Into<MetricValue>) -> Self {
+        Self {
+            key: key.into(),
+            value: value.into(),
+        }
+    }
+}
+
+/// Values metric labels may assume
+#[derive(Clone, Debug, derive_more::From)]
+pub enum MetricValue {
+    String(String),
+    Int(i64),
+    Float(f64),
+    Bool(bool),
+    // can add array if needed
+}
+impl From<&'static str> for MetricValue {
+    fn from(value: &'static str) -> Self {
+        MetricValue::String(value.to_string())
+    }
+}
+
+pub trait Counter: Send + Sync {
+    fn add(&self, value: u64, attributes: &MetricAttributes);
+}
+
+pub trait Histogram: Send + Sync {
+    // When referring to durations, this value is in millis
+    fn record(&self, value: u64, attributes: &MetricAttributes);
+}
+
+pub trait Gauge: Send + Sync {
+    // When referring to durations, this value is in millis
```

Comment:
Just crawling this PR again...is there a use of gauge for durations?

Reply:
I don't think so at the moment

Reply:
:+1: I vote we remove the comment and when we do have duration based gauges (I can't think of when) we can revisit how to expose.